tests/unit/subscriptions/test-class-wc-payments-subscriptions-event-handler.php

```diff
@@ -0,0 +1,281 @@
+			->with( $mock_subscription )
+			->willReturn( null );
+
+		$this->subscriptions_event_handler->handle_invoice_upcoming( $test_body );
+	}
+
+	/**
+	 * Test handle_invoice_paid when the incoming webhook belongs to a subscription that doesn't exist.
+	 */
+	public function test_handle_invoice_paid_exception() {
+		$test_body = [
+			'data' => [
+				'object' => [
+					'subscription' => 'sub_ID_no_invoice_paid',
+					'customer'     => 'cus_test1234',
+					'id'           => 'ii_testInvoiceID',
+				],
+			],
+		];
+
+		$this->expectException( Rest_Request_Exception::class );
+		$this->expectExceptionMessage( 'Cannot find subscription for the incoming "invoice.paid" event.' );
+
+		$this->subscriptions_event_handler->handle_invoice_paid( $test_body );
+	}
+
+	/**
+	 * Test handle_invoice_paid
+	 */
+	public function test_handle_invoice_paid_parent_order() {
+		$wcpay_subscription_id = 'sub_invoiceUpcoming';
+		$wcpay_customer_id     = 'cus_test1234';
+		$wcpay_invoiceid       = 'ii_testInvoiceID';
+		$test_body             = [
+			'data' => [
+				'object' => [
+					'subscription' => $wcpay_subscription_id,
+					'customer'     => $wcpay_customer_id,
+					'id'           => $wcpay_invoiceid,
+				],
+			],
+		];
+
+		$mock_subscription = new WC_Subscription();
+		$mock_subscription->update_meta_data( self::SUBSCRIPTION_ID_META_KEY, $wcpay_subscription_id );
+		$mock_subscription->next_payment = time();
+
+		$this->expectException( Rest_Request_Exception::class );
+		$this->expectExceptionMessage( 'Cannot find subscription for the incoming "invoice.paid" event.' );
+
+		$this->subscriptions_event_handler->handle_invoice_paid( $test_body );
+	}
+
+	/**
+	 * Test handle_invoice_paid
+	 */
+	public function test_handle_invoice_paid() {
+		$wcpay_subscription_id = 'sub_invoiceUpcoming';
+		$wcpay_customer_id     = 'cus_test1234';
+		$wcpay_invoiceid       = 'ii_testInvoiceID';
+		$test_body             = [
+			'data' => [
+				'object' => [
+					'subscription' => $wcpay_subscription_id,
+					'customer'     => $wcpay_customer_id,
+					'id'           => $wcpay_invoiceid,
+				],
+			],
+		];
+
+		$mock_renewal_order = WC_Helper_Order::create_order();
+		$mock_subscription  = new WC_Subscription();
+
+		WC_Subscriptions::set_wcs_get_subscription(
+			function ( $id ) use ( $mock_subscription ) {
+				return $mock_subscription;
+			}
+		);
+
+		WC_Subscriptions::wcs_create_renewal_order(
+			function ( $subscription ) use ( $mock_renewal_order ) {
+				return $mock_renewal_order;
+			}
+		);
+
+		$this->mock_invoice_service->expects( $this->once() )
+			->method( 'set_order_invoice_id' )
+			->with( $mock_renewal_order, $wcpay_invoiceid )
+			->willReturn( null );
+
+		$this->mock_invoice_service->expects( $this->once() )
+			->method( 'mark_pending_invoice_paid_for_subscription' )
+			->with( $mock_subscription )
+			->willReturn( null );
+
+		$this->subscriptions_event_handler->handle_invoice_paid( $test_body );
+	}
```

Comment:
Can we also get tests for `invoice.payment_failed`?

Reply:
Whoops! Yeah I forgot about this 😂 
Thanks for pointing that out

Reply:
Added in 5c835b8